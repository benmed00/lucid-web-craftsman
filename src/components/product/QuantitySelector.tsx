import React from 'react';

interface QuantitySelectorProps {
  quantity: number;
  onIncrement: () => void;
  onDecrement: () => void;
  readOnly?: boolean;
}

const QuantitySelector: React.FC<QuantitySelectorProps> = ({ 
  quantity, 
  onIncrement, 
  onDecrement, 
  readOnly = false 
}) => {
  return (
    <div>
      <label className="block text-sm font-medium text-stone-700 mb-2">
        Quantit√©
      </label>
      <div className="flex">
        <button
          className="border border-stone-300 rounded-l-md px-3 py-2 hover:bg-stone-50"
          onClick={onDecrement}
          disabled={readOnly || quantity <= 1} // Disable if quantity is 1 or less, or if readOnly
        >
          -
        </button>
        <input
          type="text"
          value={quantity}
          readOnly // Input is always readOnly, controlled by buttons
          className="border-t border-b border-stone-300 px-4 py-2 w-16 text-center focus:outline-none"
        />
        <button
          className="border border-stone-300 rounded-r-md px-3 py-2 hover:bg-stone-50"
          onClick={onIncrement}
          disabled={readOnly}
        >
          +
        </button>
      </div>
    </div>
  );
};

export default QuantitySelector;
