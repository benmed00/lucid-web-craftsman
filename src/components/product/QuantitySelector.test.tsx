// src/components/product/QuantitySelector.test.tsx
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import QuantitySelector from './QuantitySelector'; // Assuming path is correct

describe('QuantitySelector', () => {
  const mockOnIncrement = vi.fn();
  const mockOnDecrement = vi.fn();

  beforeEach(() => {
    mockOnIncrement.mockClear();
    mockOnDecrement.mockClear();
  });

  it('renders initial quantity correctly', () => {
    render(
      <QuantitySelector
        quantity={5}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );
    // The quantity is displayed in an input field
    expect(screen.getByDisplayValue('5')).toBeInTheDocument();
  });

  it('calls onIncrement when "+" button is clicked', async () => {
    const user = userEvent.setup();
    render(
      <QuantitySelector
        quantity={5}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );

    const incrementButton = screen.getByRole('button', { name: '+' });
    await user.click(incrementButton);

    expect(mockOnIncrement).toHaveBeenCalledTimes(1);
  });

  it('calls onDecrement when "-" button is clicked', async () => {
    const user = userEvent.setup();
    render(
      <QuantitySelector
        quantity={5} // Start with a quantity > 1 so decrement is enabled
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );

    const decrementButton = screen.getByRole('button', { name: '-' });
    await user.click(decrementButton);

    expect(mockOnDecrement).toHaveBeenCalledTimes(1);
  });

  it('disables "-" button when quantity is 1', () => {
    render(
      <QuantitySelector
        quantity={1}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );
    const decrementButton = screen.getByRole('button', { name: '-' });
    expect(decrementButton).toBeDisabled();
  });

  it('does not call onDecrement when quantity is 1 and "-" button is clicked (though disabled)', async () => {
    const user = userEvent.setup();
    render(
      <QuantitySelector
        quantity={1}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );
    const decrementButton = screen.getByRole('button', { name: '-' });
    if (!decrementButton.hasAttribute('disabled')) { // Should be disabled, but guard click
        await user.click(decrementButton);
    }
    expect(mockOnDecrement).not.toHaveBeenCalled();
  });

  it('disables buttons when readOnly is true', () => {
    render(
      <QuantitySelector
        quantity={5}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
        readOnly={true}
      />
    );
    const incrementButton = screen.getByRole('button', { name: '+' });
    const decrementButton = screen.getByRole('button', { name: '-' });

    expect(incrementButton).toBeDisabled();
    expect(decrementButton).toBeDisabled(); // Also disabled because readOnly
  });

  it('matches snapshot', () => {
    const { container } = render(
      <QuantitySelector
        quantity={3}
        onIncrement={mockOnIncrement}
        onDecrement={mockOnDecrement}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
