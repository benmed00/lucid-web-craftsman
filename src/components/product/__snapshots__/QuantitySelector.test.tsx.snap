// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuantitySelector > matches snapshot 1`] = `
<div>
  <div>
    <label
      class="block text-sm font-medium text-stone-700 mb-2"
    >
      Quantit√©
    </label>
    <div
      class="flex"
    >
      <button
        class="border border-stone-300 rounded-l-md px-3 py-2 hover:bg-stone-50"
      >
        -
      </button>
      <input
        class="border-t border-b border-stone-300 px-4 py-2 w-16 text-center focus:outline-none"
        readonly=""
        type="text"
        value="3"
      />
      <button
        class="border border-stone-300 rounded-r-md px-3 py-2 hover:bg-stone-50"
      >
        +
      </button>
    </div>
  </div>
</div>
`;
